### PR TITLE
allow supporting HEAD request

### PR DESCRIPTION
**Pull request by svoronuk**
For head requests headers are added to returning response

**Main problem**: headers from fetch return [Headers object](https://developer.mozilla.org/en-US/docs/Web/API/Headers) which is iterator. It has plenty of methods for iterate, but all of them do not supported by IE.

**What's next**: iterator by itself has various possibilities for handling values, but non of them supported by IE. However i found in repository two methods which pretty widely used and can be used also for handling iterator:
1. `[...iterator]`
2. `new Map(iterator)`

I used first variant in current approach

**Doubts**: I not 100% sure that files where i found spread syntax is used by CG, so may be "IE constraint" is still actual for my approach

### CODE REVIEW

File: src/core/fetchRequest.js [hunk 1]
```diff
@@ -102,6 +102,14 @@ const requestFactory = (url, options) => {
 
             // successful request
             if ((responseCode >= 200 && responseCode < 300) || (response && response.success === true)) {
+                if(options.method === "HEAD"){
+                    const headers = {};
+                    for (const [key, value] of originalResponse.headers.entries()) {
+                        headers[key] = value;
+                    }
+                    return Object.assign(response, {headers});
+                }
```
Comment (jsconan) [inline src/core/fetchRequest.js:111]: **nitpick:** `entries()` is not supported by [all browsers](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries#browser_compatibility) we must support, and it is not polyfilled.

**thought:** Is it actually needed to apply such a loop to copy the object? Wouldn't a shallow copy of the object be enough?

**suggestion:** Consider using a more classical approach, for instance something like:
```suggestion
                if(options.method === "HEAD"){
                    const headers = Object.assign({}, originalResponse.headers);
                    return Object.assign(response, { headers });
                }
```

Or even, if the object does not need to be cloned:
```suggestion
                if(options.method === "HEAD"){
                    const { headers } = originalResponse;
                    return Object.assign(response, { headers });
                }
```

**Review by jsconan — CHANGES_REQUESTED** [review]
Quick code review so far. While the main idea looks ok, the implementation needs some change as explained in the comment below.

File: src/core/fetchRequest.js (hunk 1 above)
Comment (svoronuk) [reply src/core/fetchRequest.js]: Good catch about supporting! I forget that Curr Gen supports wider version of browsers.

**Review by jsconan — COMMENTED** [review]
The code looks good. Any chance to have it unit tested?

File: src/core/fetchRequest.js (hunk 1 above)
Comment (svoronuk) [reply src/core/fetchRequest.js]: I created one more approach for headers collection. I've updated PR description with more info about situation.

**Review by jsconan — APPROVED** [review]
Very good job!

**Review Checklist**
- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [x] Pull request's target is not `master`
- [x] Commits are following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] Commits messages are meaningful
- [x] Commits are atomic
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)

**Review by olga-kulish — APPROVED** [review]
Works well with related `tao-deliver-fe` PR, and unit tests run well.
Spread operator isn't supported by IE, but I see it already used in the same repository https://github.com/oat-sa/tao-core-sdk-fe/blob/master/src/core/digest.js#L45 so I guess it's fine and covered by @babel/preset-env.

**Comment by krampstudio** [comment]
> Works well with related `tao-deliver-fe` PR, and unit tests run well.
> Spread operator isn't supported by IE, but I see it already used in the same repository https://github.com/oat-sa/tao-core-sdk-fe/blob/master/src/core/digest.js#L45 so I guess it's fine and covered by @babel/preset-env.

yes, babel will transpile it

**Review by krampstudio — COMMENTED** [review]
I'd like to challenge this PR.
For me it's not necessary. 
The whole `fetch` API isn't supported by IE. It's the reason why we use the fetch polyfill if necessary that contains the `Headers` support as well.

**Comment by oatymart** [comment]
@krampstudio without this part the following flow occurs in my other PR:
- make HEAD request from [here](https://github.com/oat-sa/tao-deliver-fe/pull/33/files#diff-79cc01a889045da0919fb71618afdd830187f5743c8a50931ab649caa1994862R182)
- response has `{ status: 200 }`, headers, and no body
- `response.json()` throws error parsing empty body
- caught by catch block which returns `{}` as response
- now the response in the next block has no `.headers` so we need to re-attach `originalResponse.headers`

I don't understand what you mean about the `fetch` polyfill. It doesn't replace this post-fetch handling code.

**Comment by jsconan** [comment]
@krampstudio For what I understood, the purpose of this PR is to transform the `Headers` into a regular object and place it inside the response for some reason. It is made after `fetch` has been called.

**Comment by krampstudio** [comment]
ok, my bad I thought the headers were already in the returned object as a `Headers` object.

It leads to another question: wouldn't it be more simple to return the `Response` object when there's no response JSON response body? 

We are doing an exception for `HEAD` but we could want to access headers also for `OPTIONS` or any method without a JSON response (`GET` to request a text file or an image).

Can't we have an option that allow to return the `Response` or fallback on the `Response` when it's not JSON
```js
if (returnOriginalResponse || response.headers.get("content-type").indexOf("application/json") !== -1) {
   return response;
}
return response.json().catch(() => ({}));
``` 

so even on usual request, we could have access to response headers

**Comment by oatymart** [comment]
> Can't we have an option that allow to return the `Response` or fallback on the `Response` when it's not JSON

Yes it seems reasonable - I'll get to work on that.

**Comment by oatymart** [comment]
I've implemented the proposed `options.returnOriginalResponse` feature. Didn't want to get into the details of checking `content-type` for responses, since I don't need it, and doing that can potentially change the behaviour and have unintended consequences for other consumers.

**Review by krampstudio — APPROVED** [review]
:+1:  

- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [ ] <s>New code is not subject to concurrency issues (if applicable)</s>
- [x] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [x] Pull request title and description are meaningful

**Pull request merged**